### DISCUSSION TITLE
bug: Client component does not receive changed data after refetch, how to resolve?

### DISCUSSION THREAD

**Discussion by ridhwaans** [trpc/trpc#6166] [Q&A] [answered]
### Provide environment information

```

  System:
    OS: Linux 4.19 Ubuntu 24.04.1 LTS 24.04.1 LTS (Noble Numbat)
    CPU: (16) x64 AMD Ryzen 7 3700X 8-Core Processor
    Memory: 1.90 GB / 7.73 GB
    Container: Yes
    Shell: 5.9 - /usr/bin/zsh
  Binaries:
    Node: 20.13.1 - ~/.nvm/versions/node/v20.13.1/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 10.5.2 - ~/.nvm/versions/node/v20.13.1/bin/npm
    pnpm: 9.1.0 - ~/.local/share/pnpm/pnpm
    bun: 1.1.27 - ~/.bun/bin/bun
  npmPackages:
    typescript: ^5.4.5 => 5.6.3
```

### Describe the bug

```javascript
'use client';

import React, {
  createContext,
  useState,
  useContext,
  ReactNode,
  useEffect,
} from 'react';

import { api } from '@/lib/trpc/react';
import { useUserContext } from '@/contexts/UserContext';

interface LibraryContextType {
  libraryItems: Array<Record<string, any>>;
  recentlyWatchedItems: Array<Record<string, any>>;
  recentlyPlayedItems: Array<Record<string, any>>;
  addItemToCollection: (item: any) => Promise<void>;
  removeItemFromCollection: (item: any) => Promise<void>;
  removeCollection: (item: any) => Promise<void>;
}

const LibraryContext = createContext<LibraryContextType | undefined>(undefined);

export const LibraryProvider: React.FC<{ children: ReactNode }> = ({
  children,
}) => {
  const { data: libraryItems, refetch: refetchLibraryItems } =
    api.home.queryAll.useQuery(
      {
        filterBy: { collection: 'library' },
      },
      {
        refetchOnWindowFocus: false,
      },
    );

  const removeCollection = async (item: any) => {
    try {
      await removeCollectionMutation.mutateAsync(item);
    } catch (error) {
      console.error('Failed to remove collection:', error);
    }
  };

        const removeCollectionMutation = api.library.removeCollection.useMutation({
          onSuccess: async (newItem) => {
            console.log('Remove collection success:', newItem);
            if (newItem.collection === 'library') {
              await refetchLibraryItems();
            }
          },
          onError: (error) => {
            console.error('Error removing item:', error);
          },
        });
      
        return (
          <LibraryContext.Provider
            value={{
              libraryItems,
              removeItemFromCollection,
              removeCollection,
            }}
          >
            {children}
          </LibraryContext.Provider>
        );
      };
      
      export const useLibraryContext = () => {
        const context = useContext(LibraryContext);
        if (context === undefined) {
          throw new Error('useLibrary must be used within a LibraryProvider');
        }
        return context;
      };
```

I have a libraryItems state and a context that I use in the client component below

```javascript
'use client';

      export const ClientCarousel = ({
        options,
        title,
        subtitle,
        className = 'md:basis-1/2 lg:basis-1/3',
      }: CarouselProps) => {
        const { libraryItems } =
          useLibraryContext();
      
        const { user } = useUserContext();
        const removeCollectionMutation = api.library.removeCollection.useMutation();
        const handleRemoveCollection = async (collection) => {
          await removeCollectionMutation.mutateAsync({
            profileId: user?.id as string,
            collection: collection,
          });
        };
        let data;
        let isLoading = false;

        if (options?.filterBy?.collection === 'library') {
          data = libraryItems;
        } else {
          const { data: queryData, isLoading: queryLoading } =
            api.home.queryAll.useQuery(options);
      
          data = queryData;
          isLoading = queryLoading;
        }
        console.log(`libraryItems refetched is this a re-render`);
        console.log(`libraryItems length is ${libraryItems?.length}`);
        return (
          <> ... display libraryItems data...  </>
        )
```

 
 I have an add item to collection mutation also that actually works
 When add mutation happens the same client component re-renders, calls api.home.queryAll.useQuery(options) and actually shows a new item 
        However the remove collection mutation does not work even though it uses the same patterns, but it does not call the api.home.queryAll.useQuery 

Is it not re-rendering with the updated data because the refetch happens in the LibraryProvider and not in the current component?
  Do I need a useeffect? I do not want to use setstate variables because i prefer data get set directly from trpc and I already have a usecontext hook above

When I hard reload the page, the client component shows the new data

### Link to reproduction

https://stackblitz.com/github/trpc/examples-next-minimal-starter?file=README.md

### To reproduce

 I put log statements to debug and saw the following: 
1) there is 1 item in the library collection  
2) removeCollection mutation is performed, library collection is deleted so there should be 0 items in the client side
3)  It goes to `onSuccess` and prints `console.log('Remove collection success:'`
4) client component re-renders and prints log statement but libraryItems length is still 1, expected 0

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/6165">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/6165/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/trpc/trpc/issues/6165/pledge.svg">
</picture>
</a>

**Comment by KATT** [comment]
I don't see the bug - converting to discussion.

**Reply by ridhwaans** [reply]
```javascript
const { libraryItems } = useLibraryContext();
```
This data is getting piped into the client component from the provider

When `removeCollectionMutation.mutateAsync` happens, it refetches the libraryItems which has changed
The issue is that the client component is not receiving the new libraryItems, or, has received it but is not re-rendering the component with the new data

**Reply by KATT** [reply, marked as answer]
https://trpc.io/docs/client/react/useUtils#query-invalidation

**Reply by ridhwaans** [reply]
I tried that, but it didnt work. I am probably doing it wrong, sorry

```javascript
const { data: libraryItems, refetch: refetchLibraryItems } =
    api.home.queryAll.useQuery(
      {
        filterBy: { collection: 'library' },
      },
      {
        refetchOnWindowFocus: false,
      },
    );  
  ```
Doesnt `refetch` update data faster and changes are immediately visible, whereas `invalidate` marks data as stale & refetches on the next access?  

Also it shouldnt matter if the mutate method lives in the context provider and passed down to the client component or is directly in the client component

using `"@tanstack/react-query": "^5.39.0",` btw (weird, i dont see refetch in react query https://tanstack.com/query/v5/docs/framework/react/reference/useQuery )

**Reply by ridhwaans** [reply]
nvm, it works now 
invalidate takes several seconds but it does re-render the client, whereas refetch felt faster
I think I need to make it optimistically show the final state while the mutation completes
thank you
